File: Intranet/test_api/java/ru/yandex/intranet/d/web/api/folders/FoldersApiTest.java
package ru.yandex.intranet.d.web.api.folders;

import java.util.Set;
import java.util.stream.Collectors;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.ParameterizedTypeReference;
import org.springframework.http.MediaType;
import org.springframework.test.web.reactive.server.WebTestClient;

import ru.yandex.intranet.d.IntegrationTest;
import ru.yandex.intranet.d.TestFolders;
import ru.yandex.intranet.d.TestUsers;
import ru.yandex.intranet.d.web.MockUser;
import ru.yandex.intranet.d.web.model.ErrorCollectionDto;
import ru.yandex.intranet.d.web.model.PageDto;
import ru.yandex.intranet.d.web.model.folders.FolderDto;

/**
 * Folders public API test.
 *
 * @author Dmitriy Timashov <dm-tim@yandex-team.ru>
 */
@IntegrationTest
public class FoldersApiTest {

    @Autowired
    private WebTestClient webClient;

    @Test
    public void getFolderTest() {
        FolderDto result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders/{id}", "aa6a5d64-5b94-4057-8d43-e65812475e73")
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(FolderDto.class)
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertEquals("aa6a5d64-5b94-4057-8d43-e65812475e73", result.getId());
    }

    @Test
    public void getFolderNotFoundTest() {
        ErrorCollectionDto result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders/{id}", "12345678-9012-3456-7890-123456789012")
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isNotFound()
                .expectBody(ErrorCollectionDto.class)
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertFalse(result.getErrors().isEmpty());
    }

    @Test
    public void getFoldersPageTest() {
        PageDto<FolderDto> result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders")
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertFalse(result.getItems().isEmpty());
    }

    @Test
    public void getFoldersTwoPagesTest() {
        PageDto<FolderDto> firstResult = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders?limit={limit}", 1)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(firstResult);
        Assertions.assertEquals(1, firstResult.getItems().size());
        Assertions.assertTrue(firstResult.getNextPageToken().isPresent());
        PageDto<FolderDto> secondResult = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders?limit={limit}&pageToken={token}",
                        1, firstResult.getNextPageToken().get())
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(secondResult);
        Assertions.assertFalse(secondResult.getItems().isEmpty());
    }

    @Test
    public void getServiceFoldersPageTest() {
        PageDto<FolderDto> result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/services/{id}/folders", 1)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertFalse(result.getItems().isEmpty());
    }

    @Test
    public void getServiceFoldersNotFoundPageTest() {
        ErrorCollectionDto result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/services/{id}/folders", 9999)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isNotFound()
                .expectBody(ErrorCollectionDto.class)
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertFalse(result.getErrors().isEmpty());
    }

    @Test
    public void getServiceFoldersTwoPagesTest() {
        PageDto<FolderDto> firstResult = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/services/{id}/folders?limit={limit}", 1, 1)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(firstResult);
        Assertions.assertEquals(1, firstResult.getItems().size());
        Assertions.assertEquals("11d1bcdb-3edc-4c21-8a79-4570e3c09c21", firstResult.getItems().get(0).getId());
        Assertions.assertTrue(firstResult.getNextPageToken().isPresent());

        PageDto<FolderDto> secondResult = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/services/{id}/folders?limit={limit}&pageToken={token}",
                        1, 1, firstResult.getNextPageToken().get())
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(secondResult);
        Assertions.assertFalse(secondResult.getItems().isEmpty());
        Assertions.assertEquals("5f73c44f-7569-43d6-980d-7ee663c9789d", secondResult.getItems().get(0).getId());
    }

    @Test
    public void getFolderAcceptableForDAdminsTest() {
        FolderDto result = webClient
                .mutateWith(MockUser.uid(TestUsers.D_ADMIN_UID))
                .get()
                .uri("/api/v1/folders/{id}", "aa6a5d64-5b94-4057-8d43-e65812475e73")
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(FolderDto.class)
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertEquals("aa6a5d64-5b94-4057-8d43-e65812475e73", result.getId());
    }

    @Test
    public void getServiceFoldersPageAcceptableForDAdminsTest() {
        PageDto<FolderDto> result = webClient
                .mutateWith(MockUser.uid(TestUsers.D_ADMIN_UID))
                .get()
                .uri("/api/v1/services/{id}/folders", 1)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertFalse(result.getItems().isEmpty());
    }

    @Test
    public void getFoldersByIdsTest() {
        PageDto<FolderDto> result = webClient
                .mutateWith(MockUser.uid(TestUsers.USER_1_UID))
                .get()
                .uri("/api/v1/folders/_by_ids?ids={id1},{id2}", TestFolders.TEST_FOLDER_2_ID,
                        TestFolders.TEST_FOLDER_4_ID)
                .accept(MediaType.APPLICATION_JSON)
                .exchange()
                .expectStatus()
                .isOk()
                .expectBody(new ParameterizedTypeReference<PageDto<FolderDto>>() { })
                .returnResult()
                .getResponseBody();
        Assertions.assertNotNull(result);
        Assertions.assertEquals(2, result.getItems().size());
        Set<String> ids = result.getItems().stream().map(FolderDto::getId).collect(Collectors.toSet());
        Assertions.assertEquals(Set.of(TestFolders.TEST_FOLDER_2_ID, TestFolders.TEST_FOLDER_4_ID), ids);
    }

}

File: Intranet/test_api/java/ru/yandex/intranet/d/web/admin/folders/FoldersApiTest.kt
package ru.yandex.intranet.d.web.admin.folders

import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.test.web.reactive.server.WebTestClient
import ru.yandex.intranet.d.IntegrationTest
import ru.yandex.intranet.d.TestFolders.*
import ru.yandex.intranet.d.TestProviders
import ru.yandex.intranet.d.TestServices.*
import ru.yandex.intranet.d.TestUsers.*
import ru.yandex.intranet.d.model.folders.FolderType.COMMON
import ru.yandex.intranet.d.util.Uuids
import ru.yandex.intranet.d.web.MockUser
import ru.yandex.intranet.d.web.model.ErrorCollectionDto
import ru.yandex.intranet.d.web.model.folders.FrontFolderDto
import ru.yandex.intranet.d.web.model.folders.FrontFolderInputDto
import ru.yandex.intranet.d.web.model.providers.FullProviderDto

@IntegrationTest
class FoldersApiTest(
    @Autowired val webClient: WebTestClient,
) {
    @Test
    fun onlyForDAdminsTest() {
        val createResponse = webClient
            .mutateWith(MockUser.uid(NOT_D_ADMIN_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_a", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .is4xxClientError
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!
        assertEquals(1, createResponse.errors.size)
        assertTrue(createResponse.errors.contains("Access denied."))

        val updateResponse = webClient
            .mutateWith(MockUser.uid(NOT_D_ADMIN_UID))
            .put()
            .uri("/admin/folders/{id}", TEST_FOLDER_1_SERVICE_ID)
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_a", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .is4xxClientError
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!
        assertEquals(1, updateResponse.errors.size)
        assertTrue(updateResponse.errors.contains("Access denied."))
    }

    @Test
    fun createFolderTest() {
        val folder = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_a", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FrontFolderDto::class.java)
            .returnResult()
            .responseBody!!
        assertNotNull(folder)
        assertEquals(COMMON, folder.folderType)
        assertTrue(Uuids.isValidUuid(folder.id))
    }

    @Test
    fun createFolderTestWrongTag() {
        val errorCollection = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_A", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isEqualTo(HttpStatus.UNPROCESSABLE_ENTITY)
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!

        assertNotNull(errorCollection)
        val errors = errorCollection.fieldErrors
        assertTrue(errors.containsKey("tags"))
        val details = errorCollection.details
        val tagsDetails = details["tags"]!!
        assertEquals(1, tagsDetails.size)
        val tagValues = tagsDetails.stream().findFirst().orElseThrow() as Map<*, *>
        assertEquals("tag_A", tagValues["wrong_value"])
        assertEquals("tag_a", tagValues["suggest_value"])
    }

    @Test
    fun createFolderInClosingServiceTest() {
        val errorCollection = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_SERVICE_ID_CLOSING,
                    "Folder name",
                    "Folder description",
                    setOf()
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isNotFound
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!

        assertNotNull(errorCollection)
        assertNotNull(errorCollection.fieldErrors)
        val errors = errorCollection.fieldErrors["serviceId"]!!
        assertTrue(errors.contains("Current service status is not allowed."))

    }

    @Test
    fun createFolderInNonExportableServiceTest() {
        val errorCollection = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_SERVICE_ID_NON_EXPORTABLE,
                    "Folder name",
                    "Folder description",
                    setOf()
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isNotFound
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!

        assertNotNull(errorCollection)
        assertNotNull(errorCollection.fieldErrors)
        val errors = errorCollection.fieldErrors["serviceId"]!!
        assertNotNull(errors)
        assertTrue(errors.contains("Services in the sandbox are not allowed."))
    }

    @Test
    fun createFolderInRenamingServiceTest() {
        webClient
            .mutateWith(MockUser.uid(USER_1_UID))
            .post()
            .uri("/admin/folders")
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_SERVICE_ID_RENAMING,
                    "Folder name",
                    "Folder description",
                    setOf()
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
    }

    @Test
    fun updateFolderTest() {
        val folder = webClient
            .mutateWith(MockUser.uid(USER_1_UID))
            .put()
            .uri("/admin/folders/{id}", TEST_FOLDER_1_ID)
            .bodyValue(
                FrontFolderInputDto(
                    0,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_a", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FrontFolderDto::class.java)
            .returnResult()
            .responseBody!!

        assertNotNull(folder)
        assertEquals(COMMON, folder.folderType)
        assertEquals(1, folder.version)
    }

    @Test
    fun updateFolderTestWrongVersion() {
        val errorCollection = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .put()
            .uri("/admin/folders/{id}", TEST_FOLDER_1_ID)
            .bodyValue(
                FrontFolderInputDto(
                    1,
                    TEST_FOLDER_1_SERVICE_ID,
                    "New folder name",
                    "New folder description",
                    setOf("tag_a", "tag_b")
                )
            )
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isEqualTo(HttpStatus.PRECONDITION_FAILED)
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!

        assertNotNull(errorCollection)
        val error = errorCollection.fieldErrors["version"]!!.stream().findFirst()
        assertTrue(error.isPresent)
        assertEquals("Version mismatch.", error.get())

    }

    @Test
    fun deleteFolderTest() {
        val expected = FrontFolderDto.from(
            TEST_FOLDER_1.toBuilder()
                .setDeleted(true)
                .setVersion(1)
                .setNextOpLogOrder(1L)
                .build()
        )
        webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .delete()
            .uri("/admin/folders/{id}", TEST_FOLDER_1_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FrontFolderDto::class.java)
            .isEqualTo(expected)
    }

    @Test
    fun deleteReserveFolderTest() {
        val provider = webClient
            .mutateWith(MockUser.uid(USER_1_UID))
            .get()
            .uri("/admin/providers/{id}", TestProviders.YP_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FullProviderDto::class.java)
            .returnResult()
            .responseBody
        assertNotNull(provider)
        assertEquals(provider!!.reserveFolderId, TEST_FOLDER_1_RESERVE_ID)

        webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .delete()
            .uri("/admin/folders/{id}", TEST_FOLDER_1_RESERVE_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FrontFolderDto::class.java)
            .returnResult()

        val result = webClient
            .mutateWith(MockUser.uid(USER_1_UID))
            .get()
            .uri("/admin/providers/{id}", TestProviders.YP_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isOk
            .expectBody(FullProviderDto::class.java)
            .returnResult()
            .responseBody
        assertNotNull(result)
        assertNull(result!!.reserveFolderId)

        webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .get()
            .uri("/front/folders/{id}", TEST_FOLDER_1_RESERVE_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isNotFound
    }

    @Test
    fun deleteFolderDefaultTest() {
        val errorCollection = webClient
            .mutateWith(MockUser.uid(D_ADMIN_UID))
            .delete()
            .uri("/admin/folders/{id}", TEST_FOLDER_2_ID)
            .accept(MediaType.APPLICATION_JSON)
            .exchange()
            .expectStatus()
            .isEqualTo(HttpStatus.LOCKED)
            .expectBody(ErrorCollectionDto::class.java)
            .returnResult()
            .responseBody!!

        val error = errorCollection.errors.stream().findFirst()
        assertTrue(error.isPresent)
        assertEquals("Cannot delete default folder.", error.get())
    }
}
